core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -352,7 +353,12 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
               // this is an optimization: if the partitions are already in the metadata reply OK immediately
               Left(Errors.NONE)
             } else {
-              Right(coordinatorEpoch, txnMetadata.prepareAddPartitions(partitions.toSet, time.milliseconds()))
+              // If verifyOnly, we should have returned in the step above. If we didn't the partitions are not present in the transaction.
+              if (verifyOnly) {
+                Left(Errors.INVALID_TXN_STATE)
```

Comment:
Suppose that some of the partitions are added to the transaction and some are not. We return `INVALID_TXN_STATE` which means the broker cannot distinguish which partitions have been added correctly. Does it matter? I think there are two main consequences:
1. We have to reject the full `Produce` request, which may contain writes to multiple partitions. This means the producer also cannot tell which partition was not added correctly. Maybe this is fine since the producer must abort anyway?
2. The `AddPartitionsToTxn` cannot batch across multiple `Produce` requests from the same `transactionalId`. Or, if it does batch across requests, then we would have to return INVALID_TXN_STATE in all of the responses. Maybe this is also fine for the same reason?

I do think we'll want to have at least in the logs somewhere which partitions were not present in the transaction.

Reply:
You bring up good points and I can give more detailed response on partition level.

Reply:
I think this was mostly an oversight on my part.

Reply:
+1 here, since we do have the finest error-code level at the per-partition level we can leverage it to be more detailed.

As for whether we should also add a global error code: in general I'd like that idea, and I think there are a few error codes: auth, unsupported_version, unsupported_format (we also have a `UNSUPPORTED_FOR_MESSAGE_FORMAT` for that), that should always be global to save the sender time to dig into lower-leveled ones. And then for different request type there are some global errors based on just the global fields --- for this case, the only global field left now is `verifyOnly` so that we probably do not have additional error codes. I vaguely remember there's a KIP proposing such a principle but I cannot find the KIP now.

If we agree this is a generally good idea then let's start with this one first :P

Reply:
Ack -- we may make verifyOnly not a global field, but the point still stands.